docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,447 @@
+column-type change to be in progress for any given table at a time.
+
+## Conversion classes
+
+Certain classes of type conversions lend themselves to faster or
+less-complicated approaches and are described herein. A full table of
+type conversions is [described later](#allowable-conversions) in this
+RFC.
+
+### Trivial
+
+Trivial conversions are those that involve no conversion of on-disk data
+and which cannot create any conflict between existing data and the
+updated schema, such as converting `STRING -> BYTES` or `STRING(10) ->
+STRING(25)`. These enable a fast-path conversion by simply updating the
+table descriptor.
+
+### Validated
+
+Validated conversions involve no conversion of on-disk data, but do
+require that all existing data is validated against the proposed schema
+change before completion. This would include conversions such as
+`STRING(100) -> STRING(10)` or `BYTES -> STRING`.
+
+In cases where existing data conflicts with the requested type, the type
+conversion will fail, rather than attempting to automatically coerce
+non-conforming data. In the example of `STRING(100) -> STRING(10)`
+above, the user would have the option of using an assisted cast
+(described below) to achieve the desired result e.g. `USING
+substring(my_string_column, 1, 10)`.
+
+These have a cost somewhere between trivial and general conversions,
+since on-disk data does not need to be rewritten, only revalidated.
+
+### Cast
+
+Cast conversions involve applying a built-in cast behavior to the
+existing column data requiring the on-disk KV data to be rewritten, such
+as `UUID -> STRING`.
+
+### Assisted
+
+Assisted conversions involve applying a `USING expression` to the
+existing column data, such as `FLOAT -> INT USING ceil(column_name)`.
+
+## Indexed columns
+
+Columns that are part of an index are supported, however columns that
+are part of a primary key are explicitly out of scope for this change.
+
+## Two-phase change
+
+Given that a type-change process may take an arbitrary and unpredictable
+amount of time to complete, we support an optional two-phase approach to
+making the schema change visible to SQL clients. This allows the user to
+choose exactly when the change to the table schema will become visible,
+allowing application downtime to be minimized in non-trivial-change
+cases.
+
+The `ALTER TABLE` command is first run with the `DEFERRED` option, which
+begins the type-conversion in the background. At any point in time, but
+generally after the job has completed, the user may issue the same
+`ALTER TABLE` command, replacing `DEFERRED` with `COMPLETE` to allow the
+type change to become visible once the job has completed. The `ABORT`
+option allows an ongoing type-conversion operation to be canceled.
+
+```
+-- Convert a STRING to the 2.0 INET type.
+-- This command will return a row containing a job id.
+ALTER TABLE rpc_logs
+  ALTER COLUMN caller_ip
+  SET DATA TYPE INET
+  DEFERRED
```

Comment:
I think this command should block until the the entire type conversion has happened (with backfill), preparing the table for the quick execution of COMPLETE below.